File: scripts/stable-launcher.js
const fs = require("fs");
const os = require("os");
const path = require("path");

// The one path that never changes: every external registration (Cursor, Explorer, Start menu, startup entry) points here, and every build repoints it at the newest exe — so registrations never go stale.
const BAT = path.join(os.homedir(), "AppData", "Local", "Termiot", "termiot.bat");

function writeStableBat(exePath) {
    fs.mkdirSync(path.dirname(BAT), { recursive: true });
    fs.writeFileSync(BAT, `@echo off\r\nstart "" "${exePath}" %*\r\n`);
    console.log(`Stable launcher -> ${exePath}`);
}

module.exports = { writeStableBat, BAT };

File: Termiot/StableLauncher.cs
using System.IO;
using System.Text.RegularExpressions;

namespace Termiot;

// The version-stable entry point: %LOCALAPPDATA%\Termiot\termiot.bat forwards to whatever exe it currently names. All external registrations point at the bat, builds repoint it (scripts/stable-launcher.js), and EnsureNewest self-heals it at startup — so nothing ever references a stale instance folder.
public static class StableLauncher
{
    public static string BatPath => Path.Combine(AppPaths.Root, "termiot.bat");

    private static readonly Regex TargetRegex = new("^start \"\" \"([^\"]+)\"", RegexOptions.Compiled);

    public static string? CurrentTarget()
    {
        try
        {
            if (!File.Exists(BatPath))
            {
                return null;
            }
            foreach (var line in File.ReadAllLines(BatPath))
            {
                var match = TargetRegex.Match(line.Trim());
                if (match.Success)
                {
                    return match.Groups[1].Value;
                }
            }
        }
        catch
        {
        }
        return null;
    }

    public static void WriteForThisExe()
    {
        var exe = Environment.ProcessPath!;
        File.WriteAllText(BatPath, $"@echo off\r\nstart \"\" \"{exe}\" %*\r\n");
        AppLog.Write("ui", "stable launcher -> " + exe);
    }

    // Newest build wins: repoint only when the bat is missing/broken or this exe is newer than its current target, so an old lingering instance can't downgrade it.
    public static void EnsureNewest()
    {
        try
        {
            var exe = Environment.ProcessPath!;
            var target = CurrentTarget();
            if (target != null && string.Equals(target, exe, StringComparison.OrdinalIgnoreCase))
            {
                return;
            }
            if (target == null || !File.Exists(target) || File.GetLastWriteTimeUtc(exe) > File.GetLastWriteTimeUtc(target))
            {
                WriteForThisExe();
            }
        }
        catch (Exception ex)
        {
            AppLog.Write("ui", "stable launcher update failed: " + ex.Message);
        }
    }
}
